### PR TITLE
Refactor manifold classes

### PR DESCRIPTION
**Pull request by nkoep**
See https://github.com/pymanopt/pymanopt/commit/f37e21ec4213e51e2c25663a824a551f9c856708 for motivation/details.

### CODE REVIEW

**Comment by nkoep** [comment]
I think this is good to go now. All tests pass, and all examples (few as they are) run on my machine.

File: pymanopt/manifolds/sphere.py [hunk 1]
```diff
@@ -103,70 +84,102 @@ def _normalize(self, X):
...
+    """Manifold of shape n1 x n2 x ... x nk tensors with unit 2-norm. The
+    metric is such that the sphere is a Riemannian submanifold of Euclidean
+    space.
 
-        if self.dim == 0:
-            warnings.warn("Manifold only consists of isolated points when "
-                          "subspace is 1-dimensional.")
+    Notes
+    -----
+    The implementation of the Weingarten map is taken from [1]_.
```
Comment (sweichwald) [inline pymanopt/manifolds/sphere.py]: Typo? (trailing underscore)

Comment (nkoep) [reply pymanopt/manifolds/sphere.py]: This is the way footnotes are defined in rst (see [here](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#footnotes)). Our docstrings right now are very inconsistent in general. I'm planning to clean this up after the backend rewrite by adopting the numpy docstring format.

Comment (sweichwald) [reply pymanopt/manifolds/sphere.py]: Thanks!

**Comment by nkoep** [comment]
> Impressive and clean refactoring @nkoep !
> 
> (Shall we collect TODOs somehow? One idea would be to raise issues for TODOs and maybe tag them "help wanted" / "good first issue" much like in sklearn (https://scikit-learn.org/stable/developers/contributing.html#issues-for-new-contributors). This way there would be some central book keeping without the overhead to keep an redundant todo file up to date.)

That's a very good idea to streamline the process a bit more. I have a feeling that items in a `TODO.md` file go there to die most of the time ;)

**Pull request merged**